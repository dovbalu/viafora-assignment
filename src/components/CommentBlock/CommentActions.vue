<template>
  <div class="flex items-center md:w-full lg:w-2/5 mt-2 space-x-4">
    <TextLabelWithHandlers
      :colorVariant="colorVariant"
      @click="handleReply()"
      class="cursor-pointer"
      data-test="reply"
      >reply</TextLabelWithHandlers
    >
    <TextLabelWithHandlers :colorVariant="colorVariant"
      >{{ replies }} replies</TextLabelWithHandlers
    >
    <!--    upvote/downvote: making chevron AND text clickable vs. only chevron clickable, for the sake of mobile users where using only chevron may be too small-->
    <TextLabelWithHandlers
      :colorVariant="colorVariant"
      @click="handleUpvote"
      class="cursor-pointer"
      data-test="upvote"
      ><span class="chevron up">⌃</span>{{ upvotes }}</TextLabelWithHandlers
    >
    <TextLabelWithHandlers
      :colorVariant="colorVariant"
      @click="handleDownvote"
      class="cursor-pointer"
      data-test="downvote"
      ><span class="chevron down">⌃</span>
      {{ downvotes }}</TextLabelWithHandlers
    >
  </div>
</template>

<script>
import TextLabelWithHandlers from "../common/TextLabelWithHandlers";
export default {
  components: { TextLabelWithHandlers },
  props: {
    colorVariant: {
      type: String,
      required: false,
      default: "light"
    },
    commentStats: {
      type: Object,
      required: true,
      default: () => ({
        upvotes: 0,
        downvotes: 0,
        replies: 0
      }),
      note: `
            {
                replies: Number,
                upvotes: Number,
                downvotes: Number,
            }
            `
    }
  },
  data() {
    const { upvotes = 0, downvotes = 0, replies = 0 } = this.commentStats;
    return {
      upvotes,
      downvotes,
      replies
    };
  },
  methods: {
    handleReply() {
      this.$emit("reply");
    },
    handleUpvote() {
      this.$emit("upvote");
    },
    handleDownvote() {
      this.$emit("downvote");
    }
  }
};
</script>
<style>
.chevron {
  display: inline-block;
}
.chevron.down {
  display: inline-block;
  @apply transform;
  @apply rotate-180;
}
</style>
