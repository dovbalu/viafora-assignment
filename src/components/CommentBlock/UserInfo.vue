<template>
  <div class="flex items-end space-x-2">
    <TextLabelWithHandlers fontVariant="bold">{{
      fullName
    }}</TextLabelWithHandlers>
    <!--    didnt do it here, but would make a separate common component for the pill-->
    <div
      class="rounded-full py-0 px-4 bg-gray-800 text-xs text-white uppercase"
    >
      {{ role }}
    </div>
  </div>
</template>

<script>
import TextLabelWithHandlers from "@/components/common/TextLabelWithHandlers";
export default {
  components: { TextLabelWithHandlers },
  props: {
    fullName: {
      type: String,
      required: true,
      default: "Jon Doe",
      note: "user full name"
    },
    role: {
      type: String,
      required: false,
      note: "user title"
    }
  }
};
</script>
