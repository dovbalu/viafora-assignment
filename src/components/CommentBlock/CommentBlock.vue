<template>
  <div>
    <div
      class="comment-block border-b-2 border-gray-300 hover:darker"
      v-if="fullName && comment"
      :key="id"
      @mouseover="colorVariant = 'dark'"
      @mouseleave="colorVariant = 'light'"
    >
      <div
        class="flex p-4"
        @mouseover="hover = true"
        @mouseleave="hover = false"
      >
        <Avatar :fullName="fullName" :avatar="avatar" />
        <div class="flex flex-col ">
          <div class="flex space-x-1">
            <UserInfo :fullName="fullName" :role="role" />
            <TextLabelWithHandlers colorVariant="light" v-if="timePosted">
              &#8226; {{ timePosted }}</TextLabelWithHandlers
            >
          </div>
          <div class="mt-2">
            <TextLabelWithHandlers data-test="comment-content">{{ text }}</TextLabelWithHandlers>
            <CommentActions
              :commentStats="comment"
              :colorVariant="colorVariant"
              @reply="handleReply"
              @upvote="handleUpvote"
              @downvote="handleDownvote"
            />
          </div>
        </div>
      </div>
    </div>
    <div v-else>Loading...</div>
  </div>
</template>

<script>
import UserInfo from "@/components/CommentBlock/UserInfo";
import CommentActions from "@/components/CommentBlock/CommentActions";
import TextLabelWithHandlers from "@/components/common/TextLabelWithHandlers";
import Avatar from "@/components/common/Avatar";
import moment from "moment";

export default {
  name: "CommentBlock",
  components: { UserInfo, TextLabelWithHandlers, Avatar, CommentActions },
  props: {
    commentData: {
      type: Object,
      required: true,
      default: () => ({
        user: {},
        comment: {}
      }),
      note: `
            {
                comment: Object,
                user: Object
            }
            `
    }
  },
  data() {
    const {
      comment: { text = "", timePosted, id } = {},
      comment,
      user: { fullName, avatar, role } = {}
    } = this.commentData || {};
    return {
      text,
      fullName,
      avatar,
      role,
      //in real life i'd transform the data in Vuex or parent component and wouldn't ask CommentBlock to do it :)
      timePosted: moment.unix(timePosted).fromNow(),
      id,
      colorVariant: "light",
      comment
    };
  },
  methods: {
    handleReply() {
      alert(`reply on comment id=${this.id}`);
    },
    handleUpvote() {
       alert(`upvote comment id=${this.id}`);
    },
    handleDownvote() {
      alert(`downvote comment id=${this.id}`);
    }
  }
};
</script>
<style>
.comment-block:first-child {
  @apply border-t-2;
}
</style>
