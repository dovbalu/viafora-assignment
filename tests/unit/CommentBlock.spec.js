import { mount } from "@vue/test-utils";
import Vue from 'vue';
import CommentBlock from "@/components/CommentBlock/CommentBlock.vue";
import commentData from "@/data";
import CommentActions from "@/components/CommentBlock/CommentActions";

describe("CommentBlock.vue", () => {
  let view = null;
  const mock = jest.fn();
  const createMockComponent = props => {
    jest.mock("moment", () => () => ({ fromNow: () => "a day ago" }));
    return mount(CommentBlock, {
      propsData: { commentData: props }
    });
  };

  beforeEach(() => {
    window.alert = mock;
    view = createMockComponent(commentData);
  });
  afterEach(() => {
    jest.clearAllMocks();
  });

  describe("renders CommentBlock", () => {
    it("renders CommentBlock correctly when correct props are provided", () => {
      expect(view).toMatchSnapshot();
    });
    it("renders 'Loading...' when props are missing", () => {
      view = createMockComponent({...commentData, user: {}});
      expect(view).toMatchSnapshot();
    });
    it("renders default image for avatar if not provided", () => {
      view = createMockComponent({...commentData, user: {...commentData.user, avatar: undefined}});
      expect(view.find('img').attributes('src')).toBe('@/assets/default.svg');
    });
  });
  describe("UI interactions - ", () => {
    const testMethod = (method, selector) => {
      const spy = jest.spyOn(CommentBlock.methods, method);
      view = createMockComponent(commentData);
      view.find(`[data-test="${selector}"]`).trigger("click");
      expect(spy).toHaveBeenCalled();
    };
    it("click on reply triggers reply method", () => {
      testMethod("handleReply", "reply");
    });
    it("click on upvote triggers handleUpvote method", () => {
      testMethod("handleUpvote", "upvote");
    });
    it("click on downvote triggers correct action", () => {
      testMethod("handleDownvote", "downvote");
    });

    it("mouseover/mouseleave causes reply block color scheme to change", async () => {
      const mouseOutScheme = view.findComponent(CommentActions).html();
      view.find('.comment-block').trigger("mouseover");
       await Vue.nextTick();
      const mouseOverScheme = view.findComponent(CommentActions).html();
      expect(mouseOutScheme).not.toEqual(mouseOverScheme);
      view.find('.comment-block').trigger("mouseleave");
      await Vue.nextTick();
      const mouseOutScheme2 = view.findComponent(CommentActions).html();
      expect(mouseOutScheme).toEqual(mouseOutScheme2);
    });
  });
});
