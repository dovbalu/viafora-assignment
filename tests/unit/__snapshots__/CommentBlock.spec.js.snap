// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentBlock.vue renders CommentBlock renders 'Loading...' when props are missing 1`] = `
<div>
  <div>Loading...</div>
</div>
`;

exports[`CommentBlock.vue renders CommentBlock renders CommentBlock correctly when correct props are provided 1`] = `
<div>
  <div class="comment-block border-b-2 border-gray-300 hover:darker">
    <div class="flex p-4">
      <div class="w-20 mr-4">
        <div class="group w-full rounded-full overflow-hidden  cursor-pointer"><span class="hidden group-hover:table-cell text-white font-bold align-middle">Brad</span> <img src="https://pickaface.net/gallery/avatar/unr_random_180410_1905_z1exb.png" alt="Brad" class="object-cover object-center w-full h-full visible group-hover:hidden"></div>
      </div>
      <div class="flex flex-col ">
        <div class="flex space-x-1">
          <div class="flex items-end space-x-2"><span class="text-sm font-bold text-black ">Brad</span>
            <div class="rounded-full py-0 px-4 bg-gray-800 text-xs text-white uppercase">
              Author
            </div>
          </div> <span class="text-sm font-normal text-gray-500 ">
            • a day ago</span>
        </div>
        <div class="mt-2"><span class="text-sm font-normal text-black " data-test="comment-content">So what the German automaker is likely to focus on today is the bigger picture. This will be the first time we see the Taycan free from any prototype bodywork.</span>
          <div class="flex items-center md:w-full lg:w-2/5 mt-2 space-x-4"><span class="cursor-pointer text-sm font-normal text-gray-500 " data-test="reply">reply</span> <span class="text-sm font-normal text-gray-500 ">12 replies</span> <span class="cursor-pointer text-sm font-normal text-gray-500 " data-test="upvote"><span class="chevron up">⌃</span>123</span> <span class="cursor-pointer text-sm font-normal text-gray-500 " data-test="downvote"><span class="chevron down">⌃</span>
            0</span></div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
